File: frontend/src/api/room.js
import { message } from 'antd';

const API_URL = 'http://localhost:8000/api/v1/room';

export const createRoom = async (access_token, roomData) => {
  try {
    const response = await fetch(`${API_URL}/create`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      },
      body: JSON.stringify({
        access_token: access_token,
        data: roomData
      })
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Ошибка создания комнаты');
    }
    
    const data = await response.json();
    message.success('Комната успешно создана!');
    return data;
  } catch (error) {
    message.error(error.message);
    return null;
  }
};

export const updateRoom = async (access_token, room_uuid, updateData) => {
  try {
    const response = await fetch(`${API_URL}/update`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      },
      body: JSON.stringify({
        room_uuid: room_uuid,
        access_token: access_token,
        update_data: updateData
      })
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Ошибка обновления комнаты');
    }
    
    const data = await response.json();
    message.success('Комната успешно обновлена!');
    return data;
  } catch (error) {
    message.error(error.message);
    return null;
  }
};

export const getRoomInfo = async (access_token, room_uuid) => {
  try {
    const response = await fetch(`${API_URL}/info/${room_uuid}?access_token=${access_token}`, {
      method: 'GET',
      headers: {
        'accept': 'application/json'
      }
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Ошибка получения информации о комнате');
    }
    
    return await response.json();
  } catch (error) {
    message.error(error.message);
    return null;
  }
};

export const fetchUserRooms = async (access_token) => {
  try {
    const response = await fetch(
      `${API_URL}/my?access_token=${access_token}`,
      {
        headers: {
          'accept': 'application/json'
        }
      }
    );
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Ошибка загрузки комнат');
    }
    
    const data = await response.json();    
    return data;
  } catch (error) {
    message.error(error.message);
    return { 
      rooms: []
    };
  }
};

export const createInviteLink = async (access_token, room_uuid, permissions) => {
  try {
    const response = await fetch(`${API_URL}/invite_link`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      },
      body: JSON.stringify({
        access_token: access_token,
        room_uuid: room_uuid,
        permissions: permissions
      })
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Ошибка создания ссылки-приглашения');
    }
    
    const data = await response.json();
    message.success('Ссылка-приглашение создана!');
    return data;
  } catch (error) {
    message.error(error.message);
    return null;
  }
};

export const connectToRoom = async (access_token, invite_link, room_password) => {
  try {
    const response = await fetch(`${API_URL}/connect/${invite_link}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      },
      body: JSON.stringify({
        access_token: access_token,
        room_password: room_password
      })
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Ошибка подключения к комнате');
    }
    
    const data = await response.json();
    message.success('Успешное подключение к комнате!');
    return data;
  } catch (error) {
    message.error(error.message);
    return null;
  }
};



export const connectToRoomByInvite = async (token, inviteLink, roomPassword) => {
  const response = await fetch(`${API_URL}/connect/${inviteLink}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({
      access_token: token,
      room_password: roomPassword,
    }),
  });

  if (!response.ok) {
    throw new Error('Ошибка подключения');
  }
  
  const data = await response.json();
  localStorage.setItem('room_permissions', data.permissions);
  return data.room_uuid;
};


export const checkRoomAccess = async (token, roomId) => {
  try {
    const response = await fetch(`${API_URL}/access`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      },
      body: JSON.stringify({
        access_token: token,
        room_uuid: roomId
      })
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Ошибка создания комнаты');
    }
    
    const data = await response.json();
    return data;
  } catch (error) {
    message.error(error.message);
    return null;
  }
};


export const removeInvitedUser = async (access_token, room_uuid, username) => {
  try {
    const response = await fetch(`${API_URL}/invited`, {
      method: 'DELETE',
      headers: {
        'Content-Type': 'application/json',
        'accept': 'application/json'
      },
      body: JSON.stringify({
        access_token,
        room_uuid,
        username
      })
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      throw new Error(errorData.message || 'Ошибка удаления пользователя');
    }
    
    return await response.json();
  } catch (error) {
    message.error(error.message);
    return null;
  }
};

File: backend/api/v1/endpoints/room.py
from typing import Optional
from fastapi import APIRouter, Depends
from starlette.responses import JSONResponse
from models.api.room import (
    RoomRemoveInvited,
    RoomUpdateRequest,
    RoomCreateRequest,
    RoomInviteLinkRequest,
    RoomConnectingRequest,
    RoomInviteLinkResponse, RoomAccessRequest
)
from services.collaboration import Collaboration
from services.core.room_service import RoomService
from services.redis_client import get_redis_client
from utils.jwt_manager import decode_access_token
from api.v1.endpoints.websocket import collaboration


router = APIRouter(
    prefix="/room",
    tags=["room"]
)


@router.post(path="/create")
async def create_room(request: RoomCreateRequest):
    token_data = decode_access_token(
        token=request.access_token
    )

    new_room = await RoomService.create_room(
        room_init_data=request.data,
        user_id=token_data.get("user_id")
    )

    return new_room


@router.patch(path="/update")
async def update_room(request: RoomUpdateRequest):
    token_data = decode_access_token(
        token=request.access_token
    )

    room_uuid = request.room_uuid

    existing_room = await RoomService.get_by_room_uuid(
        room_uuid=room_uuid,
        owner_id=token_data.get("user_id")
    )

    if not existing_room: return "not room"

    if existing_room.owner_id != token_data.get("user_id"): return "permission"

    is_updated = await RoomService.update_room_settings(
        room_uuid=room_uuid,
        update_data=request.update_data
    )

    return is_updated


@router.post(path="/invite_link", response_class=JSONResponse, response_model=RoomInviteLinkResponse)
async def get_invite_link(
        request: RoomInviteLinkRequest,
        redis_client = Depends(get_redis_client)
) -> str | RoomInviteLinkResponse:
    token_data = decode_access_token(
        token=request.access_token
    )

    room_uuid = request.room_uuid

    existing_room = await RoomService.get_by_room_uuid(
        room_uuid=room_uuid,
        owner_id=token_data.get("user_id")
    )

    if not existing_room: return "not room"

    if existing_room.owner_id != token_data.get("user_id"): return "permission"

    invite_link, room_password, permissions = await RoomService.save_room_link(
        redis_client=redis_client,
        room_uuid=room_uuid,
        permissions=request.permissions
    )

    return RoomInviteLinkResponse(
        invite_link=invite_link,
        room_password=room_password,
        permissions=permissions
    )

@router.post(path="/connect/{invite_link}", response_class=JSONResponse)
async def connecting_room(
        invite_link: str,
        request: RoomConnectingRequest,
        redis_client = Depends(get_redis_client)
) -> Optional[str]:
    token_data = decode_access_token(
        token=request.access_token
    )

    if not token_data.get("user_id"): 
        return JSONResponse({"error": "permission"}, status_code=403)

    stored_value = await redis_client.get_value(invite_link)
    if not stored_value:
        return JSONResponse({"error": "invalid link"}, status_code=404)

    parts = stored_value.split(":")
    if len(parts) < 4:
        return JSONResponse({"error": "invalid data format"}, status_code=400)

    room_uuid, password, permissions, _ = parts

    if password != request.room_password:
        return JSONResponse({"error": "invalid password"}, status_code=403)

    await RoomService.set_invited_user(
        redis_client=redis_client,
        invited_user_id=token_data.get("user_id"),
        room_uuid=room_uuid,
        permissions=permissions
    )

    return JSONResponse({
        "room_uuid": room_uuid,
        "permissions": permissions
    })


@router.get("/info/{room_uuid}")
async def get_room_info(room_uuid: str, access_token: str):
    token_data = decode_access_token(
        token=access_token
    )

    if not token_data: return 401

    user_id = token_data.get("user_id")

    response = await RoomService.get_by_room_uuid(
        room_uuid=room_uuid,
        owner_id=user_id
    )

    return response


@router.get("/my")
async def get_my_rooms(access_token: str):
    token_data = decode_access_token(
        token=access_token
    )

    if not token_data: return 401

    user_id = token_data.get("user_id")

    response = await RoomService.get_all_by_owner(
        owner_id=user_id
    )

    return response


@router.post(path="/access")
async def check_access(
        request: RoomAccessRequest,
        redis_client = Depends(get_redis_client)
):
    return await Collaboration.verify_access(
        redis_client=redis_client,
        access_token=request.access_token,
        room_uuid=request.room_uuid
    )

@router.delete(path="/invited")
async def check_access(
        request: RoomRemoveInvited,
        redis_client = Depends(get_redis_client)
):
    print(request)
    token_data = decode_access_token(
        token=request.access_token
    )

    room_uuid = request.room_uuid

    existing_room = await RoomService.get_by_room_uuid(
        room_uuid=room_uuid,
        owner_id=token_data.get("user_id")
    )

    if not existing_room: return "not room"

    user_id = await RoomService.delete_invited_user(
        redis_client=redis_client,
        username=request.username,
        room_uuid=request.room_uuid
    )
    
    await collaboration.disconnect_user(room_uuid, user_id)
    return True
